### DISCUSSION TITLE
Large files fail to push after tracking with lfs

### DISCUSSION THREAD

**Discussion by vagnozzia408** [git-lfs/git-lfs#4774] [Q&A]
I have a few files in a subdirectory of a github repository. I identified the two files that exceed GitHub's file size limit. From the main directory, I typed `git lfs track subdirectory/filename` for each file, then typed `git add .gitattributes`.

Then I tried to push to the repository, and still received the error saying that the large files were detected and I should try Git Large File Storage. Any ideas what I'm doing wrong? From everything I can tell, I'm following the directions [here](https://git-lfs.github.com/).

**Comment by bk2204** [comment]
Hey,

The likely problem is that these large file are already in the history of your project and GitHub doesn't allow any large files in the history at all.  Those steps are appropriate for new repositories, but they don't rewrite the history to remove large blobs in existing repositories.

You can use `git lfs migrate import --everything` with an appropriate `--include` argument to rewrite the history to turn all the files or file types you've specified into LFS files, and then push.  The manual page goes into some more detail about the options you can use.

**Reply by vagnozzia408** [reply]
I think I understand. (I'm pretty new to this!) So anytime I want to add a new large file to an existing repository that already uses lfs, I should take this approach?

**Reply by bk2204** [reply]
If you're adding a new large file to an existing repository, then you can run `git lfs track` _before_ adding the large file and as long as you then commit the `.gitattributes` file, you can add those large files without any problem.  If all of your large files follow a similar pattern, you can use a pattern like `git lfs track "*.bin"`.  It's okay if some files that are smaller than 100 MB end up as LFS files.

However, if the large files are added _before_ the `.gitattributes` file is set up properly, then they'll be stored as normal Git objects in the history instead of LFS objects, and then you need to run `git lfs migrate import` to fix them.

**Reply by vagnozzia408** [reply]
So if I'm working from a local repository, I should be able to:
1. Drag the file into the folder on my computer that's linked to the main branch of the GitHub repository.
2. Run `git lfs track` on that file.
3. Commit _just_ the `.gitattributes` file. (Is there something else I need to do to update this file?)
4. Then add and commit the large file.
Does that sounds like what you're describing?
Thanks for all the help!

**Reply by bk2204** [reply]
You don't have to commit the `.gitattributes` in a commit previous to the one where you add the new file, but you do need to have updated it (e.g., by running `git lfs track`) and only then add the new file using `git add` (and then commit both if you like).  However, your approach is also fine in this case and will work, since you'll have updated the `.gitattributes` file before adding the large file.

The reason is that Git uses the `.gitattributes` changes to determine whether a file should be processed via Git LFS at the point you run `git add`.  Note also that you'll need to have done `git lfs install` on your system at least once before running `git lfs track`

Hopefully that's clear, but if not, let me know and I'll try to explain it differently.

**Reply by joeflack4** [reply]
If anyone is like me and just ran `git lfs migrate import --everything` and completely screwed up their entire repo and panicked, this helped me to undo: https://github.com/git-lfs/git-lfs/issues/3032#issuecomment-502112573
